File: apps/android/app/src/main/java/com/universalsim/extender/ConnectionStore.kt
package com.universalsim.extender

import android.content.Context
import android.os.Build
import org.json.JSONArray
import org.json.JSONObject

/**
 * A remembered host, shown on the connect screen for quick reconnect. `mode` is a
 * [Mode] name so a quick-connect reuses how you last used this host. `os` is
 * learned from the host's HostInfo after connecting (empty until then).
 */
data class SavedConnection(
    val addr: String,
    val hostname: String = "",
    val os: String = "", // "windows" | "macos" | "linux" | ""
    val mode: String = Mode.CLICKER.name,
    val pin: Int = 0, // host pairing code (0 = none)
    val hidden: Boolean = false,
    val lastConnected: Long = 0L,
    val customName: String = "", // user-given friendly name ("" = use hostname)
)

/**
 * Persists saved connections in SharedPreferences as a JSON array, one entry per
 * `host:port`. The list is small, so each mutation does a simple load/edit/save.
 */
object ConnectionStore {
    private const val PREFS = "connections"
    private const val KEY = "saved"
    private const val KEY_SENSITIVITY = "trackpadSensitivity"
    private const val KEY_DEVICE_NAME = "deviceName"

    /** This device's human-readable name, sent to the host so the extra screen this
     *  phone adds is labelled (e.g. "James's phone"). Defaults to the hardware model
     *  until the user overrides it in Advanced. Matches the iOS client. */
    fun loadDeviceName(context: Context): String =
        prefs(context).getString(KEY_DEVICE_NAME, null)?.takeIf { it.isNotBlank() } ?: Build.MODEL

    fun saveDeviceName(context: Context, value: String) {
        prefs(context).edit().putString(KEY_DEVICE_NAME, value.trim()).apply()
    }

    /** Trackpad pointer-speed multiplier, persisted app-wide (not per-host). */
    fun loadSensitivity(context: Context): Float =
        prefs(context).getFloat(KEY_SENSITIVITY, 1.6f)

    fun saveSensitivity(context: Context, value: Float) {
        prefs(context).edit().putFloat(KEY_SENSITIVITY, value).apply()
    }

    fun load(context: Context): List<SavedConnection> {
        val raw = prefs(context).getString(KEY, null) ?: return emptyList()
        return runCatching {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                SavedConnection(
                    addr = o.getString("addr"),
                    hostname = o.optString("hostname"),
                    os = o.optString("os"),
                    mode = o.optString("mode", Mode.CLICKER.name),
                    pin = o.optInt("pin", 0),
                    hidden = o.optBoolean("hidden", false),
                    lastConnected = o.optLong("lastConnected", 0L),
                    customName = o.optString("customName"),
                )
            }
        }.getOrDefault(emptyList())
    }

    /** Insert or update the entry for [addr], stamping it most-recently-used. */
    fun remember(context: Context, addr: String, mode: String, pin: Int) {
        val list = load(context).toMutableList()
        val existing = list.find { it.addr == addr } ?: SavedConnection(addr)
        list.removeAll { it.addr == addr }
        list.add(existing.copy(mode = mode, pin = pin, hidden = false, lastConnected = System.currentTimeMillis()))
        save(context, list)
    }

    /** Record the host identity (OS + machine name) learned from HostInfo. */
    fun setIdentity(context: Context, addr: String, os: String, hostname: String) {
        val list = load(context)
        if (list.none { it.addr == addr }) return
        save(context, list.map { if (it.addr == addr) it.copy(os = os, hostname = hostname) else it })
    }

    fun setHidden(context: Context, addr: String, hidden: Boolean) {
        save(context, load(context).map { if (it.addr == addr) it.copy(hidden = hidden) else it })
    }

    /** Set (or clear, with "") the user's friendly name for a saved host. */
    fun setCustomName(context: Context, addr: String, name: String) {
        save(context, load(context).map { if (it.addr == addr) it.copy(customName = name.trim()) else it })
    }

    fun delete(context: Context, addr: String) {
        save(context, load(context).filterNot { it.addr == addr })
    }

    private fun save(context: Context, list: List<SavedConnection>) {
        val arr = JSONArray()
        list.forEach { c ->
            arr.put(
                JSONObject()
                    .put("addr", c.addr)
                    .put("hostname", c.hostname)
                    .put("os", c.os)
                    .put("mode", c.mode)
                    .put("pin", c.pin)
                    .put("hidden", c.hidden)
                    .put("lastConnected", c.lastConnected)
                    .put("customName", c.customName),
            )
        }
        prefs(context).edit().putString(KEY, arr.toString()).apply()
    }

    private fun prefs(context: Context) = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}

/** A glyph standing in for the host's OS, shown on saved-connection rows. */
fun deviceEmoji(os: String): String = when (os.lowercase()) {
    "windows" -> "🪟" // 🪟
    "macos", "mac" -> "🍏" // 🍎
    "linux" -> "🐧" // 🐧
    else -> "🖥️" // 🖥️
}

/** Short human label for a stored [Mode] name; falls back to the raw text for an
 *  empty/unknown mode (matches iOS's `modeLabel`). Covers all five modes — the old
 *  version rendered raw enum names for Trackpad / Second screen. */
fun modeLabel(mode: String): String =
    runCatching { Mode.valueOf(mode).label() }.getOrDefault(mode)

File: apps/ios/ScreenExtender/ConnectionStore.swift
import Foundation
import UIKit

/// A remembered host shown on the connect screen for quick reconnect. `os` and
/// `hostname` are learned from the host's HostInfo after connecting. `mode` is
/// the last-chosen mode (empty = re-ask next time).
struct SavedConnection: Codable, Identifiable {
    var addr: String
    var hostname: String = ""
    var os: String = ""     // "windows" | "macos" | "linux" | ""
    var mode: String = ""   // Mode.rawValue, empty = show picker next time
    var pin: Int = 0
    var hidden: Bool = false
    var lastConnected: Double = 0
    /// User-given friendly name for this saved host. Shown as the main label with
    /// the hostname/address in brackets when set. Defaults to "" (use hostname).
    var customName: String = ""

    var id: String { addr }
}

/// Persists saved connections in `UserDefaults` as JSON, one entry per host:port.
enum ConnectionStore {
    private static let key = "savedConnections"
    private static let sensitivityKey = "trackpadSensitivity"
    private static let deviceNameKey = "deviceDisplayName"

    static func load() -> [SavedConnection] {
        guard let data = UserDefaults.standard.data(forKey: key),
              let list = try? JSONDecoder().decode([SavedConnection].self, from: data)
        else { return [] }
        return list
    }

    /// Insert or update the entry for `addr`, stamping it most-recently-used.
    /// Pass `mode: ""` to clear the remembered mode so the picker shows next time.
    static func remember(addr: String, mode: String = "", pin: Int = 0) {
        var list = load()
        var entry = list.first { $0.addr == addr } ?? SavedConnection(addr: addr)
        entry.mode = mode
        entry.pin = pin
        entry.hidden = false
        entry.lastConnected = Date().timeIntervalSince1970
        list.removeAll { $0.addr == addr }
        list.append(entry)
        save(list)
    }

    /// Record the host identity (OS + machine name) learned from HostInfo.
    static func setIdentity(addr: String, os: String, hostname: String) {
        var list = load()
        guard let i = list.firstIndex(where: { $0.addr == addr }) else { return }
        list[i].os = os
        list[i].hostname = hostname
        save(list)
    }

    static func setHidden(addr: String, _ hidden: Bool) {
        var list = load()
        guard let i = list.firstIndex(where: { $0.addr == addr }) else { return }
        list[i].hidden = hidden
        save(list)
    }

    /// Set (or clear, with "") the user's friendly name for a saved host.
    static func setCustomName(addr: String, _ name: String) {
        var list = load()
        guard let i = list.firstIndex(where: { $0.addr == addr }) else { return }
        list[i].customName = name.trimmingCharacters(in: .whitespacesAndNewlines)
        save(list)
    }

    static func delete(addr: String) {
        save(load().filter { $0.addr != addr })
    }

    // MARK: - Trackpad sensitivity

    static func loadSensitivity() -> Float {
        let v = UserDefaults.standard.float(forKey: sensitivityKey)
        return v > 0 ? v : 1.0
    }

    static func saveSensitivity(_ value: Float) {
        UserDefaults.standard.set(value, forKey: sensitivityKey)
    }

    // MARK: - This device's name

    /// The user-set nickname for this device, or "" if never set. Empty means
    /// "use the system device name" — see `effectiveDeviceName()`.
    static func loadDeviceName() -> String {
        UserDefaults.standard.string(forKey: deviceNameKey) ?? ""
    }

    static func saveDeviceName(_ value: String) {
        UserDefaults.standard.set(value.trimmingCharacters(in: .whitespacesAndNewlines),
                                  forKey: deviceNameKey)
    }

    /// The name to advertise to a host: the user's nickname if set, else the
    /// system device name (e.g. "iPhone"; iOS 16+ hides the personalised name).
    static func effectiveDeviceName() -> String {
        let nickname = loadDeviceName()
        return nickname.isEmpty ? UIDevice.current.name : nickname
    }

    // MARK: - Private

    private static func save(_ list: [SavedConnection]) {
        if let data = try? JSONEncoder().encode(list) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }
}

/// Emoji standing in for the host OS, for saved-connection rows. Hosts are always
/// desktop machines, so this is the OS identity (form factor is implicitly a PC).
func deviceEmoji(_ os: String) -> String {
    switch os.lowercased() {
    case "macos", "mac": return "🍎"
    case "windows": return "🪟"
    case "linux": return "🐧"
    default: return "🖥️"
    }
}

/// A friendly stand-in name for a saved host when neither a custom name nor a
/// machine name is known yet — derived from the host OS (e.g. "Windows device").
/// Used for the top line of a saved-host row so it never falls back to the IP.
func deviceFallback(_ os: String) -> String {
    switch os.lowercased() {
    case "windows":      return "Windows device"
    case "macos", "mac": return "Apple device"
    case "linux":        return "Linux device"
    case "ios":          return "iOS device"
    case "android":      return "Android device"
    default:             return "Saved host"
    }
}
